src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -125,79 +125,109 @@ internal sealed class BuildServerConnection
-                    //
-                    // Example: https://github.com/dotnet/roslyn/issues/24124
-                    return new RejectedBuildResponse();
+                    var request = BuildRequest.Create(language,
+                                                      buildPaths.WorkingDirectory,
+                                                      buildPaths.TempDirectory,
+                                                      BuildProtocolConstants.GetCommitHash(),
+                                                      arguments,
+                                                      keepAlive,
+                                                      libEnvVariable);
+
+                    return await TryCompile(pipe, request, cancellationToken).ConfigureAwait(false);
                 }
+            }
 
-                if (!holdsMutex)
+            // This code uses a Mutex.WaitOne / ReleaseMutex pairing. Both of these calls must occur on the same thread 
+            // or an exception will be thrown. This code lives in a separate non-async function to help ensure this 
+            // invariant doesn't get invalidated in the future by an `await` being inserted. 
+            static Task<NamedPipeClientStream> tryConnectToServer(
+                string pipeName,
+                BuildPathsAlt buildPaths,
+                int? timeoutOverride,
+                CreateServerFunc createServerFunc,
+                CancellationToken cancellationToken)
+            {
+                var originalThreadId = Thread.CurrentThread.ManagedThreadId;
+                var clientDir = buildPaths.ClientDirectory;
+                var timeoutNewProcess = timeoutOverride ?? TimeOutMsNewProcess;
+                var timeoutExistingProcess = timeoutOverride ?? TimeOutMsExistingProcess;
+                Task<NamedPipeClientStream> pipeTask = null;
+                IServerMutex clientMutex = null;
+                try
                 {
+                    var holdsMutex = false;
                     try
                     {
-                        holdsMutex = clientMutex.TryLock(timeoutNewProcess);
+                        var clientMutexName = GetClientMutexName(pipeName);
+                        clientMutex = OpenOrCreateMutex(clientMutexName, out holdsMutex);
+                    }
+                    catch
+                    {
+                        // The Mutex constructor can throw in certain cases. One specific example is docker containers
+                        // where the /tmp directory is restricted. In those cases there is no reliable way to execute
+                        // the server and we need to fall back to the command line.
+                        //
+                        // Example: https://github.com/dotnet/roslyn/issues/24124
+                        return null;
+                    }
+
+                    if (!holdsMutex)
+                    {
+                        try
+                        {
```

Comment:
>{ [](start = 24, length = 1)

It looks like if the code in the `try` block throws an exception, we'll try to dispose of `clientMutex` in the outer `finally`, regardless of the value of `holdsMutex`.

Reply:
Correct but that's intended. If the `clientMutex` value is non-null we will call `Dispose`. That's how all `IDisposable` instances should be treated though so I'm not sure what problem you see here.